include/picongpu/plugins/xrayScattering/XrayScatteringWriter.hpp

```diff
@@ -0,0 +1,501 @@
+/* Copyright 2020 Pawel Ordyna
+ *
+ * This file is part of PIConGPU.
+ *
+ * PIConGPU is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * PIConGPU is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "picongpu/simulation_defines.hpp"
+#include "picongpu/plugins/xrayScattering/XrayScattering.hpp"
+
+#include <pmacc/assert.hpp>
+#include <pmacc/dataManagement/DataConnector.hpp>
+#include <pmacc/math/Vector.hpp>
+#include <pmacc/static_assert.hpp>
+#include <pmacc/mappings/simulation/GridController.hpp>
+
+#include <openPMD/openPMD.hpp>
+
+namespace picongpu
+{
+namespace  plugins
+{
+namespace  xrayScattering
+{
+    using namespace picongpu::plugins::xrayScattering;
+
+
+    //! Specifies plugin functioning mode. Mirrored or chunked output possible.
+    enum class OutputMemoryLayout
+    {
+        MirrorOnNodes,
+        SplitOverNodes
+    };
+
+
+    //! Specifies complex number component.
+    enum class Component
+    {
+        Real,
+        Imag
+    };
+
+
+    //! Maps a linear index to a 2D cell position vector.
+    inline std::vector< uint64_t > map2d(
+        pmacc::math::Vector< uint64_t, DIM2 >const & size,
+        uint64_t pos
+    )
+    {
+        auto const y( pos  % size.y( ) );
+        auto const x( pos / size.y( ) );
+        return std::vector< uint64_t >{ x, y };
+    }
+
+
+    //! Converts a pmacc Vector to an std::Vector.
+    template<
+        unsigned DIM,
+        typename T
+    >
+    inline std::vector< T > asStandardVector(
```

Comment:
Not a real issue, but the code is a little inconsistent with `inline` vs. `HINLINE` using both for kinda the same thing.